pkg/state/aliases.go

```diff
@@ -65,51 +67,154 @@ func (r *aliasRecord) unmarshalBinary(data []byte) error {
+	if len(data) == 0 {
+		return nil
+	}
+	rdr := bytes.NewReader(data)
+	for {
+		al, err := readStringWithUInt8Len(rdr)
+		if err != nil {
+			if !errors.Is(err, io.EOF) {
+				return err
+			}
+			break
+		}
+		r.aliases = append(r.aliases, al)
+	}
+	return nil
+}
+
+func readStringWithUInt8Len(r io.Reader) (string, error) {
+	var l [1]uint8
+	if _, err := io.ReadFull(r, l[:]); err != nil {
+		return "", err
+	}
+	b := make([]byte, l[0])
+	if _, err := io.ReadFull(r, b); err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+func (r *addressToAliasesRecord) removeIfExists(s string) bool {
+	for i, al := range r.aliases {
+		if al == s {
+			r.aliases[i] = r.aliases[len(r.aliases)-1] // replace with the last item
+			r.aliases = r.aliases[:len(r.aliases)-1]   // cut duplicate of the last item
+			return true
+		}
+	}
+	return false
+}
+
 type aliases struct {
 	db      keyvalue.IterableKeyVal
 	dbBatch keyvalue.Batch
 	hs      *historyStorage
 
 	disabled map[string]bool
 
+	scheme          proto.Scheme
 	calculateHashes bool
 	hasher          *stateHasher
 }
 
-func newAliases(db keyvalue.IterableKeyVal, dbBatch keyvalue.Batch, hs *historyStorage, calcHashes bool) *aliases {
+func newAliases(hs *historyStorage, scheme proto.Scheme, calcHashes bool) *aliases {
 	return &aliases{
-		db:              db,
-		dbBatch:         dbBatch,
+		db:              hs.db,
+		dbBatch:         hs.dbBatch,
 		hs:              hs,
 		disabled:        make(map[string]bool),
+		scheme:          scheme,
 		calculateHashes: calcHashes,
 		hasher:          newStateHasher(),
 	}
 }
 
-func (a *aliases) createAlias(aliasStr string, info *aliasInfo, blockID proto.BlockID) error {
+func (a *aliases) createAlias(aliasStr string, addr proto.WavesAddress, blockID proto.BlockID) error {
 	key := aliasKey{aliasStr}
 	keyBytes := key.bytes()
 	keyStr := string(keyBytes)
-	r := aliasRecord{*info}
+	r := aliasRecord{
+		info: aliasInfo{
+			stolen:    a.exists(aliasStr),
+			addressID: addr.ID(),
+		},
+	}
 	recordBytes, err := r.marshalBinary()
 	if err != nil {
 		return err
 	}
 	if a.calculateHashes {
 		ar := &aliasRecordForStateHashes{
-			addr:  &info.addr,
+			addr:  addr,
 			alias: []byte(aliasStr),
 		}
 		if err := a.hasher.push(keyStr, ar, blockID); err != nil {
 			return err
 		}
 	}
-	return a.hs.addNewEntry(alias, keyBytes, recordBytes, blockID)
+	if err := a.hs.addNewEntry(alias, keyBytes, recordBytes, blockID); err != nil {
+		return errors.Wrapf(err, "failed to add alias record %q for addr %q", aliasStr, addr.String())
+	}
+	return a.addOrUpdateAddressToAliasesRecord(aliasStr, addr, blockID)
+}
+
+func (a *aliases) addOrUpdateAddressToAliasesRecord(aliasStr string, addr proto.WavesAddress, blockID proto.BlockID) error {
+	var (
+		key         = addressToAliasesKey{addressID: addr.ID()}
+		recordBytes []byte
+		err         error
+	)
+	keyBytes := key.bytes()
+	// TODO: is it ok to use newestTopEntryData instead of topEntryData?
```

Comment:
Yes, it's correct to use `newest` function, because in this case you will see the aliases created by other transactions in the same block or the bunch of blocks.